### PR TITLE
add `use_karras_sigmas` to `KDPM2DiscreteScheduler` and `KDPM2AncestralDiscreteScheduler`

### PR DESCRIPTION
**Pull request by yiyixuxu**
this PR adds support for using `use_karras_sigmas` with `KDPM2DiscreteScheduler` and `KDPM2AncestralDiscreteScheduler`

fix https://github.com/huggingface/diffusers/issues/5002

### CODE REVIEW

**Comment by yiyixuxu** [comment]
@patrickvonplaten 
is there an easy way to compare the diffusers scheduler against its k-diffusion counterpart for samplers that use `BrownianTreeNoiseSampler`?  otherwise, maybe I will hardcode a little bit to make sure the noise is the same 

The outputs for `KDPM2AncestralDiscreteScheduler` are different, but they are different when not using karras sigmas, too. 

### test KDPM2DiscreteScheduler

```python
import torch
from diffusers import StableDiffusionKDiffusionPipeline, KDPM2DiscreteScheduler, KDPM2AncestralDiscreteScheduler, StableDiffusionPipeline

## diffusers 
pipe = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
pipe = pipe.to("cuda")
pipe.scheduler = KDPM2DiscreteScheduler.from_config(
    pipe.scheduler.config, use_karras_sigmas=True
)

prompt = "an astronaut riding a horse on mars"

generator = torch.Generator(device="cuda").manual_seed(seed)
image = pipe(prompt, generator=generator, num_inference_steps=20).images[0]

image.save("test1_out_d.png")

## k-diffusion
seed = 33

pipe = StableDiffusionKDiffusionPipeline.from_pretrained(
    "CompVis/stable-diffusion-v1-4"
).to("cuda")
pipe.set_scheduler("sample_dpm_2")


prompt = "an astronaut riding a horse on mars"

generator = torch.Generator(device="cuda").manual_seed(seed)
image = pipe(
    prompt, generator=generator, num_inference_steps=20, 
    use_karras_sigmas=True
).images[0]

image.save("test1_out_k.png")
```
HF diffusers           |  k-diffusion
:-------------------------:|:-------------------------:
yiyi_test_kdiffusion_kdpm_out_d | yiyi_test_kdiffusion_kdpm_out_k


```python

# test2: KDPM2AncestralDiscreteScheduler

## diffusers

seed = 33

pipe = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
pipe = pipe.to("cuda")
pipe.scheduler = KDPM2AncestralDiscreteScheduler.from_config(
    pipe.scheduler.config, use_karras_sigmas=True
)

prompt = "an astronaut riding a horse on mars"

generator = torch.Generator(device="cuda").manual_seed(seed)
image = pipe(prompt, generator=generator, num_inference_steps=20).images[0]

image.save("test2_out_d.png")

## k-diffusion

seed = 33

pipe = StableDiffusionKDiffusionPipeline.from_pretrained(
    "CompVis/stable-diffusion-v1-4"
).to("cuda")
pipe.set_scheduler("sample_dpm_2_ancestral")


prompt = "an astronaut riding a horse on mars"

generator = torch.Generator(device="cuda").manual_seed(seed)
image = pipe(
    prompt, generator=generator, num_inference_steps=20, 
    use_karras_sigmas=True
).images[0]

image.save("test2_out_k.png")
```
HF diffusers           |  k-diffusion
:-------------------------:|:-------------------------:
yiyi_test_kdiffusion_kdpm_a_out_d | yiyi_test_kdiffusion_kdpm_a_out_k

File: src/diffusers/schedulers/scheduling_k_dpm_2_ancestral_discrete.py [hunk 1]
```diff
@@ -305,6 +315,44 @@ def sigma_to_t(self, sigma):
         t = t.view(sigma.shape)
         return t
 
+    # Copied from diffusers.schedulers.scheduling_euler_discrete.EulerDiscreteScheduler._sigma_to_t
+    def _sigma_to_t(self, sigma, log_sigmas):
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_k_dpm_2_ancestral_discrete.py:319]: @patrickvonplaten 
We have a `sigma_to_t` function that does the same math as this one but is calculated with pytorch tensors (vs np here). Should we update all the calculations of sigmas in the np array so that we can just use `_sigma_to_t`?

Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_k_dpm_2_ancestral_discrete.py:319]: I think it doesn't really matter whether we use PyTorch or numpy here. Ideally we have the same `sigma_to_t` (or `_sigma_to_t` function across all schedulers though

**Comment by patrickvonplaten** [comment]
> @patrickvonplaten is there an easy way to compare the diffusers scheduler against its k-diffusion counterpart for samplers that use `BrownianTreeNoiseSampler`? otherwise, maybe I will hardcode a little bit to make sure the noise is the same
> 
> The outputs for `KDPM2AncestralDiscreteScheduler` are different, but they are different when not using karras sigmas, too.
> 
> ### test KDPM2DiscreteScheduler
> ```python
> import torch
> from diffusers import StableDiffusionKDiffusionPipeline, KDPM2DiscreteScheduler, KDPM2AncestralDiscreteScheduler, StableDiffusionPipeline
> 
> ## diffusers 
> pipe = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
> pipe = pipe.to("cuda")
> pipe.scheduler = KDPM2DiscreteScheduler.from_config(
>     pipe.scheduler.config, use_karras_sigmas=True
> )
> 
> prompt = "an astronaut riding a horse on mars"
> 
> generator = torch.Generator(device="cuda").manual_seed(seed)
> image = pipe(prompt, generator=generator, num_inference_steps=20).images[0]
> 
> image.save("test1_out_d.png")
> 
> ## k-diffusion
> seed = 33
> 
> pipe = StableDiffusionKDiffusionPipeline.from_pretrained(
>     "CompVis/stable-diffusion-v1-4"
> ).to("cuda")
> pipe.set_scheduler("sample_dpm_2")
> 
> 
> prompt = "an astronaut riding a horse on mars"
> 
> generator = torch.Generator(device="cuda").manual_seed(seed)
> image = pipe(
>     prompt, generator=generator, num_inference_steps=20, 
>     use_karras_sigmas=True
> ).images[0]
> 
> image.save("test1_out_k.png")
> ```
> 
> HF diffusers	k-diffusion
> yiyi_test_kdiffusion_kdpm_out_d	yiyi_test_kdiffusion_kdpm_out_k
> ### test2: KDPM2AncestralDiscreteScheduler
> ```python
> ## diffusers
> 
> seed = 33
> 
> pipe = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
> pipe = pipe.to("cuda")
> pipe.scheduler = KDPM2AncestralDiscreteScheduler.from_config(
>     pipe.scheduler.config, use_karras_sigmas=True
> )
> 
> prompt = "an astronaut riding a horse on mars"
> 
> generator = torch.Generator(device="cuda").manual_seed(seed)
> image = pipe(prompt, generator=generator, num_inference_steps=20).images[0]
> 
> image.save("test2_out_d.png")
> 
> ## k-diffusion
> 
> seed = 33
> 
> pipe = StableDiffusionKDiffusionPipeline.from_pretrained(
>     "CompVis/stable-diffusion-v1-4"
> ).to("cuda")
> pipe.set_scheduler("sample_dpm_2_ancestral")
> 
> 
> prompt = "an astronaut riding a horse on mars"
> 
> generator = torch.Generator(device="cuda").manual_seed(seed)
> image = pipe(
>     prompt, generator=generator, num_inference_steps=20, 
>     use_karras_sigmas=True
> ).images[0]
> 
> image.save("test2_out_k.png")
> ```
> 
> HF diffusers	k-diffusion
> yiyi_test_kdiffusion_kdpm_a_out_d	yiyi_test_kdiffusion_kdpm_a_out_k

Hmm yeah not sure if `StableDiffusionKDiffusionPipeline` doesn't work with the brownian sampler I think you have to hack into the original code for 1-to-1 comparison

**Comment by yiyixuxu** [comment]
able to get 1-to-1 matches now for `KDPM2AncestralDiscreteScheduler` with this script 
```python
import torch
from diffusers import StableDiffusionKDiffusionPipeline, KDPM2DiscreteScheduler, KDPM2AncestralDiscreteScheduler, StableDiffusionPipeline

from diffusers.utils.torch_utils import randn_tensor
from k_diffusion.sampling import get_ancestral_step, to_d
from tqdm.auto import trange, tqdm

# YiYi notes: modified `sample_dpm_2_ancestral` to use the same noise generator as in diffusers scheduler 

def default_noise_sampler(x):
    return lambda generator: randn_tensor(x.shape, dtype=x.dtype, device=x.device, generator=generator)

@torch.no_grad()
def sample_dpm_2_ancestral(model, x, sigmas, extra_args=None, callback=None, disable=None, eta=1., s_noise=1., generator=None):
    """Ancestral sampling with DPM-Solver second-order steps."""
    extra_args = {} if extra_args is None else extra_args
    noise_sampler = default_noise_sampler(x)
    s_in = x.new_ones([x.shape[0]])
    for i in trange(len(sigmas) - 1, disable=disable):
        denoised = model(x, sigmas[i] * s_in, **extra_args)
        sigma_down, sigma_up = get_ancestral_step(sigmas[i], sigmas[i + 1], eta=eta)
        if callback is not None:
            callback({'x': x, 'i': i, 'sigma': sigmas[i], 'sigma_hat': sigmas[i], 'denoised': denoised})
        d = to_d(x, sigmas[i], denoised)
        if sigma_down == 0:
            # Euler method
            dt = sigma_down - sigmas[i]
            x = x + d * dt
        else:
            # DPM-Solver-2
            sigma_mid = sigmas[i].log().lerp(sigma_down.log(), 0.5).exp()
            dt_1 = sigma_mid - sigmas[i]
            dt_2 = sigma_down - sigmas[i]
            x_2 = x + d * dt_1
            denoised_2 = model(x_2, sigma_mid * s_in, **extra_args)
            d_2 = to_d(x_2, sigma_mid, denoised_2)
            x = x + d_2 * dt_2
            noise = noise_sampler(generator)
            noise = noise_sampler(generator)
            x = x + noise * s_noise * sigma_up
    return x


# diffusers
seed = 33

pipe = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
pipe = pipe.to("cuda")
pipe.scheduler = KDPM2AncestralDiscreteScheduler.from_config(
    pipe.scheduler.config
)

prompt = "an astronaut riding a horse on mars"

generator = torch.Generator(device="cuda").manual_seed(seed)
image = pipe(prompt, generator=generator, num_inference_steps=20).images[0]

image.save("test_out_d.png")

# k-diffusion

seed = 33

pipe = StableDiffusionKDiffusionPipeline.from_pretrained(
    "CompVis/stable-diffusion-v1-4"
).to("cuda")
## use modified "sample_dpm_2_ancestral"
pipe.sampler = sample_dpm_2_ancestral


prompt = "an astronaut riding a horse on mars"

generator = torch.Generator(device="cuda").manual_seed(seed)
image = pipe(
    prompt, generator=generator, num_inference_steps=20, 
).images[0]

image.save("test_out_k.png")
```
HF diffusers           |  k-diffusion
:-------------------------:|:-------------------------:
yiyi_test_kdiffusion_2_out_d | yiyi_test_kdiffusion_2_out_k

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_k_diffusion.py [hunk 2]
```diff
@@ -609,6 +609,9 @@ def model_fn(x, t):
             noise_sampler = BrownianTreeNoiseSampler(latents, min_sigma, max_sigma, noise_sampler_seed)
             sampler_kwargs["noise_sampler"] = noise_sampler
 
+        if "generator" in inspect.signature(self.sampler).parameters:
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_k_diffusion.py:612]: Allowing to pass a `generator` to the sampler so that it is relatively easy to make 1-to-1 comparison for k-diffusion samplers that require a noise, i.e. I can still use our k-diffusion pipeline but with a modified copy of the sampler that uses same noise constructor as diffusers , e.g. this script https://github.com/huggingface/diffusers/pull/5111#issuecomment-1730237336

happy to change it back if it is too hacky

File: src/diffusers/schedulers/scheduling_k_dpm_2_ancestral_discrete.py (hunk 1 above)
Comment (yiyixuxu) [reply src/diffusers/schedulers/scheduling_k_dpm_2_ancestral_discrete.py:302]: I replaced `sigma_to_t` with `_sigma_to_t` :)

**Comment by yiyixuxu** [comment]
@patrickvonplaten we are ready for final review here:)

File: src/diffusers/schedulers/scheduling_k_dpm_2_ancestral_discrete.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_k_dpm_2_ancestral_discrete.py:302]: works for me!

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_k_diffusion.py (hunk 2 above)
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_k_diffusion.py:612]: Totally fine for me!

**Review by patrickvonplaten — APPROVED** [review]
Great job!
